In [184]:
import numpy as np

# Максимум 5 баллов - по ~0.7 баллов за задание

### 1. Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы. 
Для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [185]:
n = 100
x = np.array([[1,0,1], [2,0,2], [3,0,3], [4,4,4]])
x = np.random.randint(2, size=(n,n))

#### 1.1 Полностью векторная реализация

In [186]:
%%timeit -n1000
(np.diag(x)[np.diag(x) != 0]).prod()

12.1 µs ± 3.51 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 1.2 Реализация без векторов

In [187]:
%%timeit -n1000
res = 1
for i in range(x.shape[1]):
    if x[i][i] != 0:
        res *= x[i][i]
res

34.9 µs ± 4.76 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 1.3 Альтернативное решение

In [188]:
%%timeit -n100
res = 1
for i, el in np.ndenumerate(x):
    if i[0] == i[1] and el != 0:
        res *= el
res

2.48 ms ± 117 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


#### 1.4 Выводы

| N | Векторный | Безвекторный | Алтернативный |
| - | ------------- | ------------- | ------------- | 
| 10  | 8.24 µs  | 5.52 µs | 23.6 µs |
| 100  | 9.08 µs  | 29.2 µs  | 2.31 ms  |
| 1000  | 11.4 µs  | 307 µs | 226 ms |

Как видим, вариант без использования векторов на малом размере массива оказался немного быстрее векторного, но затем ни один из методов не был близок по скорости к векторному методу. Причем Сам векторный метод изменился совсем не сильно по скорости отработки с увелечением размера массива

### 2. Дана матрица X и два вектора одинаковой длины i и j. Построить вектор np.array([X[i[0], j[0]], X[i[1], j[1]], ... , X[i[N-1], j[N-1]]]).

In [189]:
n = 100
x = np.random.randint(0, n*n, size=(n, n))
i = np.random.randint(0, n, size=n)
j = np.random.randint(0, n, size=n)
x = np.array([[1,0,1], [2,0,2], [3,0,3], [4,5,6]])
i = [1,1,1,1,1,1,1,1,1,3]
j = [1,1,1,1,1,1,1,1,1,2]

#### 2.1 Полностью векторная реализация

In [194]:
%%timeit -n1000
X = x[i,j]

5.6 µs ± 1.17 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 2.2 Реализация без векторов

In [191]:
%%timeit -n1000
X = np.zeros(n)
for cur in range(len(i)):
    X[cur] = x[i[cur]][j[cur]]

6.47 µs ± 1.09 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 2.3 Альтернативное решение

In [192]:
%%timeit -n10000
ij = np.vstack((i, j)).T
X = x[ ij[:, 0], ij[:, 1] ]

7.41 µs ± 1.33 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### 2.4 Выводы

| N | Векторный | Безвекторный | Алтернативный |
| - | ------------- | ------------- | ------------- | 
| 10  | 3.54 µs  | 3.99 µs | 6.28 µs |
| 100  | 3.28 µs  | 3.9 µs  | 6.44 µs  |
| 1000  | 3.8 µs  | 5.22 µs | 6.34 ms |

Вариант с использованием векторов оказался немного быстрее всех других

### 3. Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество. 
Для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [195]:
n = 100
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
x = np.random.randint(10, size = n)
y = np.random.permutation(x)

#### 3.1 Полностью векторная реализация

In [196]:
%%timeit -n1000
(np.sort(y) == np.sort(x)).all()

8.25 µs ± 1.56 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 3.2 Реализация без векторов

In [197]:
%%timeit -n1000
sorted(y) == sorted(x)

27.4 µs ± 4.26 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 3.3 Альтернативное решение

In [198]:
%%timeit -n1000
y.sort() == x.sort()

1.55 µs ± 329 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 3.4 Выводы

| N | Векторный | Безвекторный | Алтернативный |
| - | ------------- | ------------- | ------------- | 
| 10  | 6.96 µs  | 4.3 µs | 573 ns |
| 100  | 6.96 µs  | 23.7 µs  | 1.07 µs  |
| 1000  | 25 µs  | 289 µs | 7.57 µs |

Альтернативный варинт с использованием векторов оказался быстрее других, в нем использовались функции от прямого объекта, в отличие от 1 варианта. Безвекторный сильно медленее

### 4. Найти максимальный элемент в векторе x среди элементов, перед которыми стоит нулевой. 
Для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [199]:
n = 100
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
x = np.random.randint(10, size = n)
x[0] = 0
# x = np.array([ 6, 2, 0, 0, 3, 5, 7])
# x = np.array([6, 2, 1, 3, 1, 1, 5, 7, 1, 0, 99])

#### 4.1 Полностью векторная реализация

In [200]:
%%timeit -n1000
mask = np.insert(x == 0, 0, False)[:-1]
# Создали маску равных 0, которую сдвинули лишним False и срезали последний
x[mask].max()

18.1 µs ± 4.75 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 4.2 Реализация без векторов

In [201]:
%%timeit -n1000
res = 0
for i in range(len(x)-1):
    if x[i] == 0:
       res = max(res, x[i+1]) 
res

14.7 µs ± 3.79 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 4.3 Альтернативное решение

In [202]:
%%timeit -n1000
mask = np.zeros(len(x) + 1, dtype=bool)
mask[1:] = (x == 0)
x[mask[:-1]].max()

9.61 µs ± 1.95 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


#### 4.4 Выводы

| N | Векторный | Безвекторный | Алтернативный |
| - | ------------- | ------------- | ------------- | 
| 10  | 15.1 µs  | 1.25 µs | 5.57 µs |
| 100  | 14.9 µs  | 11.5 µs  | 6.49 µs  |
| 1000  | 16.4 µs  | 124 µs | 6.58 µs |

Альтернативный вариант с использованием векторов оказался на удивление быстрее других, в нем в отличие от 1 варианта использовано создание массива нулей и срезы массива для вставки и отображения. В первом варианте пользовались функцией вставки.

### 5. Дан трёхмерный массив, содержащий изображение, размера (height, width, numChannels), а также вектор длины numChannels. Сложить каналы изображения с указанными весами, и вернуть результат в виде матрицы размера (height, width). Считать реальное изображение можно при помощи функции scipy.misc.imread (если изображение не в формате png, установите пакет pillow: conda install pillow). Преобразуйте цветное изображение в оттенки серого, использовав коэффициенты np.array([0.299, 0.587, 0.114]).

In [203]:
import matplotlib.pyplot as plt
import numpy as np
from imageio import v3 as iio

In [208]:
%matplotlib inline
img = iio.imread('ex.png')[:,:,:-1] # 308 * 191
qr = np.array([0.299, 0.587, 0.114])
# plt.imshow(img)
# plt.show()
img.shape

(191, 308, 3)

#### 5.1 Полностью векторная реализация

In [209]:
%%timeit -n10
img_grey = np.dot(img, qr)

# plt.imshow(img_grey, cmap='gray')
# plt.show()

1.53 ms ± 557 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


#### 5.2 Реализация без векторов

In [216]:
%%timeit -n1
img_grey = img
for i in range(img.shape[0]):
    for j in range(img.shape[1]):
        cur = 0
        for k in range(img.shape[2]):
            cur += img[i][j][k] * qr[k]
        img_grey[i][j] = cur

# plt.imshow(img_grey, cmap='gray')
# plt.show()

146 ms ± 13.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


#### 5.3 Альтернативное решение

In [215]:
%%timeit -n10
img_grey = (img * qr)
img_grey = np.vstack(np.vstack(img_grey).sum(axis=1)) 
img_grey = img_grey.reshape(img.shape[0], img.shape[1], -1)

# plt.imshow(img_grey, cmap='gray')
# plt.show()

93.8 ms ± 5.21 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [ ]:
plt.imshow(img_grey, cmap='gray')
plt.show()

#### 5.4 Выводы

| img.shape | Векторный | Безвекторный | Алтернативный |
| - | ------------- | ------------- | ------------- | 
| (191, 308, 3)  | 1.15 ms  | 139 ms | 79.1 ms |
| (908, 908, 3)  | 12.4 ms  | 1.73 s  | 1.52 s  |
| (931, 1105, 3)  | 17.6 ms  | 2.42 s | 1.65 s |

Векторный вариант работает за одинаковые порядки времени даже при изменении объема массива, и явно быстрее других методов реализации.

### 6. Реализовать кодирование длин серий (Run-length encoding). Дан вектор x. Необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить. 
Пример: x = np.array([2, 2, 2, 3, 3, 3, 5, 2, 2]). Ответ: (np.array([2, 3, 5, 2]), np.array([3, 3, 1, 2])).

In [221]:
# x = np.array([9, 2, 2, 2, 3, 3, 3, 5, 2, 2, 7])
x = np.array([2, 2, 2, 3, 3, 3, 5, 2, 2])
for i in range(7):
    x = np.concatenate([x,x])
x.shape

(1152,)

#### 6.1 Полностью векторная реализация

In [147]:
def union_repeats(x):
    split_indices = np.flatnonzero(np.diff(x, prepend=np.nan))
    # Достаем ненулевые индексы с добавлением пустого в конец. Под ними наши уникальные числа
    numbers = x[split_indices]
    segments = np.split(x, split_indices)[1:]
    counts = [len(seg) for seg in segments]
    return numbers, counts

In [222]:
%%timeit -n100
union_repeats(x)

480 µs ± 76 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


#### 6.2 Реализация без векторов

In [149]:
def union_repeats_v2(x):
    numbers = []
    counts = []
    count = 1
    for i in range(len(x)):
        if (i == len(x) - 1) or (x[i] != x[i+1]):
            numbers.append(x[i])
            counts.append(count)
            count = 1
        else:
            count += 1
    return numbers, counts

In [223]:
%%timeit -n100
union_repeats_v2(x)

400 µs ± 107 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


#### 6.3 Альтернативное решение

In [151]:
def union_repeats_v3(x):
    indices = np.where(np.diff(x) != 0)[0] + 1
    segments = np.split(x, indices)
    numbers = np.concatenate([segment[:1] for segment in segments])
    counts = np.array([len(segment) for segment in segments])
    return numbers, counts

In [224]:
%%timeit -n100
union_repeats_v3(x)

600 µs ± 113 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


#### 6.4 Выводы

| x.shape | Векторный | Безвекторный | Алтернативный |
| - | ------------- | ------------- | ------------- | 
| 36  | 48.8 µs  | 11.4 µs | 39.5 µs |
| 576  | 210 µs  | 166 µs | 264 µs |
| 1152  | 394 µs  | 319 µs  | 503 µs  |
| 9216  | 3.18 ms  | 3.1 ms | 4.25 ms |

Безвекторный вариант оказался быстрее во всех случаях, но незначительно

### 7. Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравнить с функцией scipy.spatial.distance.cdist.

##### Замечание. Можно считать, что все указанные объекты непустые (к примеру, в задаче №1 на диагонали матрицы есть ненулевые элементы).
Полезные функции NumPy: np.zeros, np.ones, np.diag, np.eye, np.arange, np.linspace, np.meshgrid,
np.random.random, np.random.randint, np.shape, np.reshape, np.transpose, np.any, np.all, np.nonzero, np.where,
np.sum, np.cumsum, np.prod, np.diff, np.min, np.max, np.minimum, np.maximum, np.argmin, np.argmax, np.unique,
np.sort, np.argsort, np.bincount, np.ravel, np.newaxis, np.dot, np.linalg.inv, np.linalg.solve.
Многие из этих функций можно использовать так: x.argmin().

In [176]:
n = 1000
# X = np.array([[1,2],[40,2],[1,2]])
# Y = np.array([[3,2],[5,2]])
# X = np.random.randint(10, size=(500,2))
# Y = np.random.randint(10, size=(20,2))
X = np.random.randint(10, size=(n,2))
Y = np.random.randint(10, size=(n,2))

#### 7.1 Полностью векторная реализация

In [178]:
%%timeit -n100
res = np.zeros((X.shape[0], Y.shape[0]))
for i, x in enumerate(X):
    for j, y in enumerate(Y):
        res[i, j] = np.linalg.norm(x - y)
res

3.75 s ± 87.9 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [179]:
%%timeit -n100
# Добавляем в конец массива, вместо обращения по индексам
res = []
for i, x in enumerate(X):
    tmp = [] 
    for j, y in enumerate(Y):
        tmp.append(np.linalg.norm(x - y))
    res.append(tmp)
res

3.84 s ± 119 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [180]:
%%timeit -n100
# Не используем enumerate
res = []
for i in range(len(X)):
    tmp = [] 
    for j in range(len(Y)):
        tmp.append(np.linalg.norm(X[i] - Y[j]))
    res.append(tmp)
res

4.06 s ± 65.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


#### 7.2 Реализация без векторов

In [182]:
%%timeit -n10
def euclidean(x,y):
    sum = 0
    for i in range(len(x)):
        sum += (x[i] - y[i]) ** 2
    return sum ** 0.5
    
res = []
for i in range(len(X)):
    tmp = [] 
    for j in range(len(Y)):
        tmp.append(euclidean(X[i], Y[j]))
    res.append(tmp)
res

2.26 s ± 45.3 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


#### 7.3 Альтернативное решение

In [183]:
%%timeit -n10
from scipy.spatial.distance import cdist
distances = cdist(X, Y, metric='euclidean')
distances

2.14 ms ± 525 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


#### 7.4 Выводы

| N | Векторный(лучший) | Безвекторный | Алтернативный |
| - | ------------- | ------------- | ------------- | 
| 10  | 361 µs  | 209 µs | 5.72 µs |
| 100  | 32.4 ms  | 22.7 ms | 36.5 µs |
| 1000  | 3.75 s  | 2.26 s  | 2.14 ms  |

Однозначно функция cdist в альтернативном варианте отрабатывает на порядки быстрее. Векторный и безвекторные варианты примерно одинаково работают, но безвекторный получился немного быстрее.

# Для каждой из задач:
(a) Написать на Python + NumPy несколько вариантов кода различной эффективности. Должно быть не
менее трёх вариантов, в том числе как минимум один полностью векторизованный вариант (без циклов), и один
вариант без векторизации (с циклами). Третий вариант решения — на ваше усмотрение, например, это может быть
наиболее хорошо читаемый способ решения, или частично векторизованный вариант. Все пункты выполнять в одном google collab, разделяя их пунктами 1., 1.1, 1.2, 1.3; 2., 2.1, 2.2, 2.3; и тд.

(b) Сравнить в IPython Notebook при помощи %timeit скорость работы на нескольких тестовых наборах
РАЗНОГО размера (минимум 3, причём отличаться размеры входных данных должны минимум на порядок. Например, массивы длины 100, 1000, 10000).

(c) Проанализировать полученные данные о скорости работы разных реализаций.

(d) Получить выводы.

Soft Deadline - 13 марта в 23:59
Hard Deadline - 20 марта в 23:59